### ISSUE TITLE
Race between kube-proxy start up and node registration

### ISSUE THREAD

**Issue by freehan** [kubernetes/kubernetes#37414]
During kube-proxy bootstrap, kube-proxy retrieves node IP by accessing node object.
https://github.com/kubernetes/kubernetes/blob/release-1.5/cmd/kube-proxy/app/server.go#L464

Because kube-proxy is a static pod, kubelet may start kube-proxy before completing self registration. Hence kube-proxy may start up without knowing the correct node IP.

**Comment by Random-Liu** [comment]
@yujuhong to confirm.

**Comment by ApsOps** [comment]
I'm hitting this as well, for the same reason - kube-proxy comes up before node registration is complete.

```
E0412 12:44:15.750069   23359 server.go:388] Can't get Node "ip-10-0-8-170", assuming iptables proxy, err: nodes "ip-10-0-8-170" not found
I0412 12:44:15.750971   23359 server.go:203] Using iptables Proxier.
W0412 12:44:15.751910   23359 server.go:436] Failed to retrieve node info: nodes "ip-10-0-8-170" not found
W0412 12:44:15.751970   23359 proxier.go:226] invalid nodeIP, initialize kube-proxy with 127.0.0.1 as nodeIP
I0412 12:44:15.751988   23359 server.go:215] Tearing down userspace rules.
I0412 12:44:15.752397   23359 healthcheck.go:119] Initializing kube-proxy health checker
I0412 12:44:15.765032   23359 conntrack.go:81] Set sysctl 'net/netfilter/nf_conntrack_max' to 262144
I0412 12:44:15.765335   23359 conntrack.go:66] Setting conntrack hashsize to 65536
I0412 12:44:15.765595   23359 conntrack.go:81] Set sysctl 'net/netfilter/nf_conntrack_tcp_timeout_established' to 86400
I0412 12:44:15.765625   23359 conntrack.go:81] Set sysctl 'net/netfilter/nf_conntrack_tcp_timeout_close_wait' to 3600
```

I'm wondering what are the consequences of this? 🤔

**Comment by yujuhong** [comment]
> @yujuhong to confirm.

I completely missed the issue :-\

kube-proxy uses the host network, so it's not blocked by kubelet registering and getting the pod CIDR. Can't kube-proxy wait and/or retry?
@freehan is the bug still valid or has it been fixed?

**Comment by freehan** [comment]
This is not fixed. We are moving kube-proxy to daemonset. I am wondering if there is some easy config to make sure kube-proxy starts up after node status becomes ready or some sort.

**Comment by yujuhong** [comment]
> This is not fixed. We are moving kube-proxy to daemonset.
yay!

> This is not fixed. We are moving kube-proxy to daemonset. I am wondering if there is some easy config to make sure kube-proxy starts up after node status becomes ready or some sort.

Is making kube-proxy blocked until node is ready not an option?

**Comment by thockin** [comment]
@mspreitzer

**Comment by thockin** [comment]
Update?

**Comment by MikeSpreitzer** [comment]
I got derailed by #43556 ; clues there would be nice.  I will get back to this today, one way or another.

**Comment by MikeSpreitzer** [comment]
I have been reading about tests, wondering where to add a test case that covers this problem.  It is not clear to me that Kubernetes has a test suite that can test race conditions in cluster creation.

**Comment by freehan** [comment]
@MikeSpreitzer 
There are some reboot/restart test.

**Comment by xiangpengzhao** [comment]
@MikeSpreitzer FYI:

https://github.com/kubernetes/kubernetes/blob/master/test/e2e/reboot.go
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/restart.go

**Comment by MikeSpreitzer** [comment]
A test case that probes this issue would be even more intimate with the way the cluster is deployed, needing to inject a controlled delay in the node startup sequence.

**Comment by bowei** [comment]
Will this make 1.7?

**Comment by bowei** [comment]
@dchen1107  move to the next milestone

**Comment by MikeSpreitzer** [comment]
BTW, `hack/local-up-cluster.sh` always makes the kube-proxy before the kubelet, so it is easy to probe this issue in manual testing.

**Comment by bowei** [comment]
Next milestone reason: it seems that there are still some conversations about this the PR fix around whether or not we want to make `kube-proxy` require the presence of `kubelet`. It seems warranted then to accept this fix in the next release.

**Comment by matchstick** [comment]
We need to address this in a patch release, but it should not block 1.7. 
@kubernetes/sig-network-misc

**Comment by jaycoon** [comment]
This is impacting one of our worker nodes as well, which was working fine and then after a restart, the API server communication was impacted (for POST requests, e.g. `logs`, `exec`). We noticed the the `invalid nodeIP` error in the kube-proxy logs.

The workaround is to detect the failure via the logs, and restart the kube-proxy pod.

**Comment by thockin** [comment]
As far as I can tell, the nodeIP is used in one particular case and nowhere else.  Specifically a service of type=LoadBalancer, which also uses LoadBalancerSourceRanges field AND specifies the node itself as a valid source range for the LB.

That's a pretty corner case.  I bet we could find a better way to express that, and eliminate that parameter and error message entirely.  I doubt it is actually causing anyone problems unless I totally misunderstand what that field is doing

**Comment by paulsubrata55** [comment]
Since the PR #74341 is merged to master, `nodeIP` is also used in `ipvs/proxier.go`. Now if this race condition happens, kube-proxy is not able to determine the correct nodeIP. This affects all the `SCTP` nodeport services with kube-proxy mode ipvs.

Reopening the issue /reopen

**Comment by danwinship** [comment]
looks like this shouldn't have gotten closed anyway
/reopen
/remove-lifecycle rotten
/lifecycle frozen

**Comment by paulsubrata55** [comment]
Can't we use the similar fix given in PR #47262.
